kernel/frame_allocator/src/lib.rs

```diff
@@ -870,8 +1030,62 @@ fn contains_any(
 /// Currently, this function adds no new frames at all if any frames within the given `frames` list
 /// overlap any existing regions at all. 
 /// TODO: handle partially-overlapping regions by extending existing regions on either end.
-fn add_reserved_region(
-    list: &mut StaticArrayRBTree<Chunk>,
+fn add_reserved_region_to_frames_list(
+    list: &mut StaticArrayRBTree<Frames<{FrameState::Unmapped}>>,
+    frames: FrameRange,
+) -> Result<FrameRange, &'static str> {
+
+    // Check whether the reserved region overlaps any existing regions.
+    match &mut list.0 {
+        Inner::Array(ref mut arr) => {
+            for chunk in arr.iter().flatten() {
+                if let Some(_overlap) = chunk.overlap(&frames) {
+                    // trace!("Failed to add reserved region {:?} due to overlap {:?} with existing chunk {:?}",
+                    //     frames, _overlap, chunk
+                    // );
+                    return Err("Failed to add reserved region that overlapped with existing reserved regions (array).");
+                }
+            }
+        }
+        Inner::RBTree(ref mut tree) => {
+            let mut cursor_mut = tree.upper_bound_mut(Bound::Included(frames.start()));
+            while let Some(chunk) = cursor_mut.get().map(|w| w.deref()) {
+                if chunk.start() > frames.end() {
+                    // We're iterating in ascending order over a sorted tree,
+                    // so we can stop looking for overlapping regions once we pass the end of the new frames to add.
+                    break;
+                }
+                if let Some(_overlap) = chunk.overlap(&frames) {
+                    // trace!("Failed to add reserved region {:?} due to overlap {:?} with existing chunk {:?}",
+                    //     frames, _overlap, chunk
+                    // );
+                    return Err("Failed to add reserved region that overlapped with existing reserved regions (RBTree).");
+                }
+                cursor_mut.move_next();
+            }
+        }
+    }
+
+    list.insert(Frames::new(
+        MemoryRegionType::Reserved,
+        frames.clone(),
+    )).map_err(|_c| "BUG: Failed to insert non-overlapping frames into list.")?;
+
+    Ok(frames)
+}
+
+/// Adds the given `frames` to the given `list` as a Region of reserved frames. 
+/// 
+/// Returns the range of **new** frames that were added to the list, 
+/// which will be a subset of the given input `frames`.
+///
+/// Currently, this function adds no new frames at all if any frames within the given `frames` list
+/// overlap any existing regions at all. 
+/// TODO: handle partially-overlapping regions by extending existing regions on either end.
+/// 
+/// TODO: combine both functions for adding reserved regions/ frames using a trait OR make one function which adds to both lists, since they are always called together.
```

Comment:
Right now I've duplicated this function for the free list containing Frames and the regions list.
Since they're always called together, should we just have one function which checks both the RESERVED_REGIONS list and the free_reserved_frames_list?